### PR TITLE
IR: preserve argument evaluation order more carefully

### PR DESCRIPTION
**Pull request by pyos**
1. receivers should be evaluated before named arguments;
 2. just because an argument has no side effects doesn't mean it is not affected by the other arguments' side effects - in that case it should still be evaluated in source order.

 #KT-47660 Fixed

### CODE REVIEW

File: compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/CallGenerator.kt [hunk 1]
```diff
@@ -383,48 +384,41 @@ class CallGenerator(statementGenerator: StatementGenerator) : StatementGenerator
...
-
         return irBlock
     }
 }
 
-fun IrExpression.hasNoSideEffects() =
+fun IrExpression.isUnchanging() =
     this is IrFunctionExpression ||
             (this is IrCallableReference<*> && dispatchReceiver == null && extensionReceiver == null) ||
             this is IrClassReference ||
```
Comment (pyos) [inline compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/CallGenerator.kt:417]: note: technically those can have side effects due to class initialization, so maybe this is wrong

**Comment by dnpetrov** [comment]
@mglukhikh please, take a look at https://github.com/JetBrains/kotlin/pull/4497/commits/f3cdf19ea2c7bdbcb63853292b2b8bf6df7653b2#diff-b0333afff787be679b651253d6311c2686a5663a7e50d651c6c50c48943dd194R295

**Comment by dnpetrov** [comment]
PSI2IR part - LGTM.

**Review by mglukhikh — CHANGES_REQUESTED** [review]
Fir2IrTextTestGenerated.testDeprecated fails in your branch. Please fix (looks like something is broken in annotation handling).

**Comment by pyos** [comment]
fixed

**Comment by mglukhikh** [comment]
Looks ok, I'll check whether bootstrap can be built with this branch.

**Review by mglukhikh — APPROVED** [review]

**Comment by dnpetrov** [comment]
1. Please, add missing generated JS tests.

2. There were some test failures, in particular:
- JsCodegenBoxTestGenerated$ArgumentOrder.testSingleSideEffect - most likely a bug in legacy JS BE, use `IGNORE_BACKEND: JS` directive 
- Fir2IrTextTestGenerated$FirProblems.testDeprecated - please, update testData accordingly.

**Comment by dnpetrov** [comment]
Nevermind, I'll fix the bureaucracy myself.

**Comment by pyos** [comment]
>Fir2IrTextTestGenerated$FirProblems.testDeprecated - please, update testData accordingly.

That should already be fixed, see above

**Comment by dnpetrov** [comment]
master (1.6-M1): https://github.com/JetBrains/kotlin/commit/3dc7b6c3eec1ceee1917563ca9e8e71a12539d08
1.5.30-M1: https://github.com/JetBrains/kotlin/commit/afc0726ad66248b8a16cdfb1a00a7af3f24cc8cb
1.5.30: https://github.com/JetBrains/kotlin/commit/52ca6bcca84a2f75819f11bf16f7279cd0fa42fe

**Pull request closed without merging**